### PR TITLE
core: implement store interface for region tree

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5825.

### What is changed and how does it work?

This PR implements the `Store` interface defined in #5820, which can be used as the storage of informers.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -189,3 +189,5 @@ require (
 // kvproto at the same time. You can run `go mod tidy` to make it replaced with go-mod style specification.
 // After the PR to kvproto is merged, remember to comment this out and run `go mod tidy`.
 // replace github.com/pingcap/kvproto => github.com/$YourPrivateRepo $YourPrivateBranch
+
+replace github.com/pingcap/kvproto => github.com/rleungx/kvproto v0.0.0-20230109095105-0fb3bc985ef7
```
Comment (nolouch) [inline go.mod:193]: Is there a related PR?

Comment (rleungx) [reply go.mod:193]: I just sent a PR, see https://github.com/pingcap/kvproto/pull/1034

**Comment by rleungx** [comment]
@binshi-bing PTAL

File: server/core/region_store.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+// Add implements the informer interface.
+func (t *RegionTree) Add(item interface{}) error {
+	return t.Update(item)
+}
+
+// Update implements the informer interface.
+func (t *RegionTree) Update(item interface{}) error { //revive:disable
+	var r *regionItem
+	regionStat := item.(*pdpb.Item).GetRegionStat()
```
Comment (binshi-bing) [inline server/core/region_store.go:31]: Usually, do we need to check the correctness of the interface conversion?

File: server/core/region_store.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+			return false
+		}
+		res = append(res, region)
+		return true
+	})
+	return res, nil
+}
+
+// GetByKey implements the informer interface.
+func (t *RegionTree) GetByKey(key string) (item interface{}, err error) {
```
Comment (binshi-bing) [inline server/core/region_store.go:79]: do we need to return "err error"?

File: server/core/region_store.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+package core
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
```
Comment (binshi-bing) [inline server/core/region_store.go:22]: Shall we define something like below then implement all the following methods as RegionStore's?
`
type RegionStore struct {
        RegionTree
}
`

**Review by binshi-bing — COMMENTED** [review]
Left some comments

**Pull request closed without merging**